jf_agent/agent_logging.py

```diff
@@ -1,50 +1,91 @@
-from datetime import datetime
 from contextlib import contextmanager
 from functools import wraps
 import logging
 import os
-import traceback
+import sys
+import urllib3
 
 '''
 Guidance on logging/printing in the agent:
 
-stdout (e.g., print() output) is for the customer/operator; it should include
- - info on run progress: tqdm output, etc.
- - errors/warnings encountered
+stdout (e.g., print() output) is for the customer/operator; it should be used when
+we have customer sensitive data that they do not want to submit to jellyfish (i.e. passwords)
 
-logger output is for Jellyfish; it should include
+logger output is for Jellyfish and customers; it should include:
  - function entry/exit timings
  - loop iteration timings/liveness indicators
  - errors/warnings encountered
 
 logger output is submitted to Jellyfish, so should only contain non-sensitive data:
 e.g., function names, iteration counts
-
-Since we generally want errors/warnings to go to BOTH stdout and the logger, we
-should generally use log_and_print() instead of logger.whatever().
 '''
 
 LOG_FILE_NAME = 'jf_agent.log'
 
+# For styling in log files, I think it's best to always use new lines even when we use
+# the special character to ignore them. Leverage always_use_newlines for this
+def _emit_helper(_self: logging.Handler, record: logging.LogRecord, always_use_newlines=False):
+    special_code = '[!n]'
+
+    msg = _self.format(record)
+    if special_code in msg:
+        msg = msg.replace('[!n]', '')
+        if always_use_newlines:
+            msg += '\n'
+    else:
+        msg += '\n'
+
+    _self.stream.write(msg)
+    _self.flush()
+
+
+class CustomStreamHandler(logging.StreamHandler):
+    """Handler that controls the writing of the newline character"""
+
+    def emit(self, record) -> None:
+        _emit_helper(self, record)
+
+
+class CustomFileHandler(logging.FileHandler):
+    """Handler that controls the writing of the newline character"""
+
+    def emit(self, record) -> None:
+        _emit_helper(self, record, always_use_newlines=True)
```

Comment:
Q: Do we need to open the stream here if it's not already open? This is the code from logging.FileHandler:

```
def emit(self, record):
        if self.stream is None:
            if self.mode != 'w' or not self._closed:
                self.stream = self._open()
        if self.stream:
            StreamHandler.emit(self, record)
```

Reply:
Yeah that's a good idea

Reply:
Done ✔️